### PR TITLE
Document layout guarantee of MaybeUninit

### PR DESCRIPTION
**Pull request by petertodd**
Good to make this explicit!

### CODE REVIEW

**Comment by petertodd** [comment]
Also added some tests for this: https://github.com/rust-lang/rust/pull/60590

File: src/libcore/mem.rs [hunk 1]
```diff
@@ -1002,6 +1002,26 @@ impl<T: ?Sized> DerefMut for ManuallyDrop<T> {
...
+/// However remember that a type *containing* a `MaybeUninit<T>` is not necessarily the same
+/// layout.  In particular, because any bit value is valid, the compiler can't apply
+/// non-zero/niche-filling optimizations:
+///
+/// ```rust
+/// # use std::mem::{MaybeUninit, size_of, align_of};
+/// assert_eq!(size_of::<Option<bool>>(), 1);
+/// assert_eq!(size_of::<Option<MaybeUninit<bool>>>(), 2);
+/// ```
+///
```
Comment (RalfJung) [inline src/libcore/mem.rs]: Could you move this down below examples? They are closely connected to the previous story about initialization invariants.

**Comment by RalfJung** [comment]
Cc @Centril

**Comment by petertodd** [comment]
You mean move the whole layout section, or just the bit about non-zero?

The latter might make sense as part of the array initialization example, e.g. why you can't initialize an [Option<MaybeUninit<T>>] or similar, and why the transmute is safe.

Though I feel like a official guarantee statement belongs in it's own section, or in the main body.

On May 7, 2019 9:45:30 AM UTC, Ralf Jung <notifications@github.com> wrote:
>RalfJung commented on this pull request.
>
>
>
>> +/// ```rust
>+/// use std::mem::{MaybeUninit, size_of, align_of};
>+/// assert_eq!(size_of::<MaybeUninit<u64>>(), size_of::<u64>());
>+/// assert_eq!(align_of::<MaybeUninit<u64>>(), size_of::<u64>());
>+/// ```
>+///
>+/// However remember that a type *containing* a `MaybeUninit<T>` is
>not necessarily the same
>+/// layout.  In particular, because any bit value is valid, the
>compiler can't apply
>+/// non-zero/niche-filling optimizations:
>+///
>+/// ```rust
>+/// # use std::mem::{MaybeUninit, size_of, align_of};
>+/// assert_eq!(size_of::<Option<bool>>(), 1);
>+/// assert_eq!(size_of::<Option<MaybeUninit<bool>>>(), 2);
>+/// ```
>+///
>
>Could you move this down below examples? They are closely connected to
>the previous story about initialization invariants.

-- 
https://petertodd.org 'peter'[:-1]@petertodd.org

**Comment by RalfJung** [comment]
I mean the hole section. I like its text and code, but I think it should be either above "initialization invariants" or (preferrably) below "examples".

**Comment by petertodd** [comment]
Sure.

Come to think of it, I think I should also mention that rust doesn't guarantee struct layouts are the same for different types - the non-zero optimization point is only relevant if you've already solved the struct layout issue in general.

On May 7, 2019 9:55:36 AM UTC, Ralf Jung <notifications@github.com> wrote:
>I mean the hold section. I like its text and code. Either above
>"initialization invariants" or (preferrably) below "examples".

-- 
https://petertodd.org 'peter'[:-1]@petertodd.org

**Comment by RalfJung** [comment]
I'd keep the discussion focused on `MaybeUninit` though.

**Comment by petertodd** [comment]
Done!

File: src/libcore/mem.rs [hunk 2]
```diff
@@ -1110,6 +1110,28 @@ impl<T: ?Sized> DerefMut for ManuallyDrop<T> {
...
 /// [ub]: ../../reference/behavior-considered-undefined.html
+///
+/// # Layout
+///
+/// `MaybeUninit<T>` is guaranteed to have the same size and alignment as `T`:
+///
+/// ```rust
+/// use std::mem::{MaybeUninit, size_of, align_of};
+/// assert_eq!(size_of::<MaybeUninit<u64>>(), size_of::<u64>());
+/// assert_eq!(align_of::<MaybeUninit<u64>>(), size_of::<u64>());
```
Comment (RalfJung) [inline src/libcore/mem.rs]:
```suggestion
/// assert_eq!(align_of::<MaybeUninit<u64>>(), align_of::<u64>());
```

Comment (petertodd) [reply src/libcore/mem.rs]: Doh! I see why running the tests didn't catch that one. :) Fixed.

**Comment by RalfJung** [comment]
Thanks a lot!

**Pull request merged**